In [3]:
%env KERAS_BACKEND = tensorflow

env: KERAS_BACKEND=tensorflow


In [4]:
import numpy as np
import matplotlib.pyplot as plt

In [7]:
from keras.datasets import mnist
(x_train, y_train) , (x_test, y_test) = mnist.load_data()

In [8]:
x_train = x_train.reshape(60000,28,28,1)
x_test = x_test.reshape(10000,28,28,1)

In [9]:
from keras.utils import np_utils
y_train = np_utils.to_categorical(y_train,10)
y_test = np_utils.to_categorical(y_test,10)

In [10]:
x_train = x_train/255
x_test = x_test/255

In [11]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

In [18]:
model = Sequential()
model.add(Conv2D(4,(5,5), padding='same',input_shape=(28,28,1)))
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(8,(5,5), padding='same',))
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(16,(5,5), padding='same',))
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Flatten())
model.add(Dense(9))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

Instructions for updating:
Colocations handled automatically by placer.


In [20]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 28, 28, 4)         104       
_________________________________________________________________
activation_1 (Activation)    (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 14, 14, 8)         808       
_________________________________________________________________
activation_2 (Activation)    (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 7, 7, 16)          3216      
__________

In [22]:
model.compile(loss='mse', optimizer='Adam', metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=100,epochs=10)

Instructions for updating:
Use tf.cast instead.
Epoch 1/10
60000/60000 [==============================] - 64s 1ms/step - loss: 0.0232 - acc: 0.8247: 3s
Epoch 2/10
60000/60000 [==============================] - 59s 980us/step - loss: 0.0080 - acc: 0.9478
Epoch 3/10
60000/60000 [==============================] - 59s 979us/step - loss: 0.0063 - acc: 0.9591
Epoch 4/10
60000/60000 [==============================] - 59s 987us/step - loss: 0.0054 - acc: 0.9650
Epoch 5/10
60000/60000 [==============================] - 59s 983us/step - loss: 0.0047 - acc: 0.9695
Epoch 6/10
60000/60000 [==============================] - 60s 997us/step - loss: 0.0043 - acc: 0.9726
Epoch 7/10
60000/60000 [==============================] - 59s 988us/step - loss: 0.0038 - acc: 0.97570s - loss: 0.0038 - acc: 
Epoch 8/10
60000/60000 [==============================] - 61s 1ms/step - loss: 0.0035 - acc: 0.9780
Epoch 9/10
60000/60000 [==============================] - 60s 1ms/step - loss: 0.0031 - acc: 0.9803
Epoch 10/10

使用上課範例，花費時間:60s/1次訓練(總花費時間:600s)

In [23]:
score = model.evaluate(x_test,y_test)
print('loss', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 6s 551us/step
loss 0.002835978928721056
正確率 0.981


減少hidden layer 增加filter數量(16,32)

In [12]:
model_1 = Sequential()
model_1.add(Conv2D(16,(5,5), padding='same',input_shape=(28,28,1)))
model_1.add(Activation('relu'))
model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Conv2D(32,(5,5), padding='same',))
model_1.add(Activation('relu'))
model_1.add(MaxPool2D(pool_size=(2,2)))


model_1.add(Flatten())
model_1.add(Dense(9))
model_1.add(Activation('relu'))
model_1.add(Dense(10))
model_1.add(Activation('softmax'))

Instructions for updating:
Colocations handled automatically by placer.


In [37]:
model_1.compile(loss='mse', optimizer='Adam', metrics=['accuracy'])
model_1.fit(x_train, y_train, batch_size=100,epochs=2)

Epoch 1/2
60000/60000 [==============================] - 97s 2ms/step - loss: 0.0050 - acc: 0.9673
Epoch 2/2
60000/60000 [==============================] - 99s 2ms/step - loss: 0.0036 - acc: 0.9771


花費時間:100s/1次訓練(較上課範例慢) 但僅訓練2次即可達到0.9818正確率(總花費時間196s)

In [38]:
score = model_1.evaluate(x_test,y_test)
print('loss', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 7s 716us/step
loss 0.0028092992554955344
正確率 0.9818


增加hidden layer 減少filter個數

In [16]:
model_2 = Sequential()
model_2.add(Conv2D(1,(5,5), padding='same',input_shape=(28,28,1)))
model_2.add(Activation('relu'))
model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(2,(5,5), padding='same',))
model_2.add(Activation('relu'))
model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(4,(5,5), padding='same',))
model_2.add(Activation('relu'))
model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(8,(5,5), padding='same',))
model_2.add(Activation('relu'))
model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Flatten())
model_2.add(Dense(9))
model_2.add(Activation('relu'))
model_2.add(Dense(10))
model_2.add(Activation('softmax'))

In [46]:
model_2.compile(loss='mse', optimizer='Adam', metrics=['accuracy'])
model_2.fit(x_train, y_train, batch_size=100,epochs=5)

Epoch 1/5
60000/60000 [==============================] - 43s 715us/step - loss: 0.0118 - acc: 0.92271s - loss: 0.
Epoch 2/5
60000/60000 [==============================] - 48s 794us/step - loss: 0.0111 - acc: 0.9267
Epoch 3/5
60000/60000 [==============================] - 48s 798us/step - loss: 0.0107 - acc: 0.9299
Epoch 4/5
60000/60000 [==============================] - 47s 786us/step - loss: 0.0103 - acc: 0.9323
Epoch 5/5
60000/60000 [==============================] - 46s 760us/step - loss: 0.0099 - acc: 0.9355


花費時間:45s/1次訓練(較上課範例快) (總花費時間:232s)

In [47]:
score = model_2.evaluate(x_test,y_test)
print('loss', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 5s 537us/step
loss 0.00912065279554663
正確率 0.9404


更改filter大小(5,5)-->(4,4)

In [41]:
model_3 = Sequential()
model_3.add(Conv2D(1,(4,4), padding='same',input_shape=(28,28,1)))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(3,3)))

model_3.add(Conv2D(2,(4,4), padding='same',))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(2,2)))

model_3.add(Conv2D(4,(4,4), padding='same',))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(2,2)))

model_3.add(Conv2D(8,(4,4), padding='same',))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(2,2)))

model_3.add(Flatten())
model_3.add(Dense(9))
model_3.add(Activation('relu'))
model_3.add(Dense(10))
model_3.add(Activation('softmax'))

In [44]:
model_3.compile(loss='mse', optimizer='Adam', metrics=['accuracy'])
model_3.fit(x_train, y_train, batch_size=100,epochs=10)

Epoch 1/10
60000/60000 [==============================] - 24s 405us/step - loss: 0.0337 - acc: 0.7525
Epoch 2/10
60000/60000 [==============================] - 27s 443us/step - loss: 0.0286 - acc: 0.8002
Epoch 3/10
60000/60000 [==============================] - 28s 465us/step - loss: 0.0249 - acc: 0.8294
Epoch 4/10
60000/60000 [==============================] - 27s 457us/step - loss: 0.0224 - acc: 0.8476
Epoch 5/10
60000/60000 [==============================] - 27s 454us/step - loss: 0.0206 - acc: 0.8606
Epoch 6/10
60000/60000 [==============================] - 30s 494us/step - loss: 0.0194 - acc: 0.8686
Epoch 7/10
60000/60000 [==============================] - 29s 491us/step - loss: 0.0185 - acc: 0.8770
Epoch 8/10
60000/60000 [==============================] - 26s 431us/step - loss: 0.0178 - acc: 0.8811
Epoch 9/10
60000/60000 [==============================] - 27s 454us/step - loss: 0.0171 - acc: 0.8862
Epoch 10/10
60000/60000 [==============================] - 28s 474us/step - loss: 

總花費時間:273s

In [45]:
score = model_3.evaluate(x_test,y_test)
print('loss', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 4s 353us/step
loss 0.016154176737181843
正確率 0.8909


### 綜合比較
model(上課範例)
model_1(減少layer增加filter)
model_2(增加layer減少filter)
model_3(更改filter大小)

model:    600s-->正確率 98.1 %
model_1:  196s-->正確率 98.18%
model_2:  232s-->正確率 94.04%
model_3:  273s-->正確率 89.09%

-->增加filter數量較經濟